clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java

```diff
@@ -0,0 +1,255 @@
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    reportingThread.interrupt();
+  }
+
+  /**
+   * Record a leader heartbeat timestamp for a given partition of a store version from a specific region.
+   *
+   * @param store the store this heartbeat is for
+   * @param version the version this heartbeat is for
+   * @param partition the partition this heartbeat is for
+   * @param region the region this heartbeat is from
+   * @param timestamp the time of this heartbeat
+   */
+  public void recordLeaderHeartbeat(String store, int version, int partition, String region, Long timestamp) {
+    recordHeartbeat(store, version, partition, region, timestamp, leaderHeartbeatTimeStamps);
+
+  }
+
+  /**
+   * Record a follower heartbeat timestamp for a given partition of a store version from a specific region.
+   *
+   * @param store the store this heartbeat is for
+   * @param version the version this heartbeat is for
+   * @param partition the partition this heartbeat is for
+   * @param region the region this heartbeat is from
+   * @param timestamp the time of this heartbeat
+   */
+  public void recordFollowerHeartbeat(String store, int version, int partition, String region, Long timestamp) {
+    recordHeartbeat(store, version, partition, region, timestamp, followerHeartbeatTimeStamps);
+  }
+
+  private void recordHeartbeat(
+      String store,
+      int version,
+      int partition,
+      String region,
+      Long timestamp,
+      Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> heartbeatTimestamps) {
+    if (heartbeatTimestamps.get(store) != null) {
+      if (heartbeatTimestamps.get(store).get(version) != null) {
+        if (heartbeatTimestamps.get(store).get(version).get(partition) != null) {
+          if (region != null) {
+            heartbeatTimestamps.get(store).get(version).get(partition).put(region, timestamp);
+          }
+        }
+      }
+    }
+  }
+
+  protected Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> getLeaderHeartbeatTimeStamps() {
+    return leaderHeartbeatTimeStamps;
+  }
+
+  protected Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> getFollowerHeartbeatTimeStamps() {
+    return followerHeartbeatTimeStamps;
+  }
+
+  protected void recordLags(
+      Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> heartbeatTimestamps,
+      ReportLagFunction lagFunction) {
+    for (Map.Entry<String, Map<Integer, Map<Integer, Map<String, Long>>>> storeName: heartbeatTimestamps.entrySet()) {
+      for (Map.Entry<Integer, Map<Integer, Map<String, Long>>> version: storeName.getValue().entrySet()) {
+        for (Map.Entry<Integer, Map<String, Long>> partition: version.getValue().entrySet()) {
+          for (Map.Entry<String, Long> region: partition.getValue().entrySet()) {
+            lagFunction.apply(storeName.getKey(), version.getKey(), region.getKey(), region.getValue());
+          }
+        }
+      }
+    }
```

Comment:
This one I'm not sure about... is the iteration safe in the presence of concurrent updates to these maps? It may be fine I suppose.

Reply:
This should be thread safe, I think the only situation where this might go sidewise is that an entry we retrieved from the iterator is no longer there, and then when we call recordLag() we just hit one of the null checks and then carry on.  Other scenario is something gets added that we don't loop through on this round of iteration, which is ok, because we'll just get it next time.